sdk/src/azure/core/az_base64.c

```diff
@@ -404,7 +404,7 @@ AZ_NODISCARD az_result
 az_base64_decode(az_span destination_bytes, az_span source_base64_text, int32_t* out_written)
 {
   _az_PRECONDITION_VALID_SPAN(destination_bytes, 1, false);
-  _az_PRECONDITION_VALID_SPAN(source_base64_text, 4, false);
+  _az_PRECONDITION_VALID_SPAN(source_base64_text, 2, false);
```

Comment:
I don't believe this is valid base64, if we are strictly adhering to the spec, which this method was meant for:
https://datatracker.ietf.org/doc/html/rfc4648#section-4

> The encoding process represents 24-bit groups of input bits as output
   strings of **4 encoded characters**.  Proceeding from left to right, a
   24-bit input group is formed by concatenating 3 8-bit input groups.
   These 24 bits are then treated as 4 concatenated 6-bit groups, each
   of which is translated into a single character in the base 64
   alphabet.

Reply:
> > With JWS or JWK's, padding characters are explicitly not allowed. We need to allow for the assumption of padding characters should the text be `length % 4 != 0`.
> 
> I don't fully understand the purpose of this PR. Do we need full Base64 **Url** encoding/decoding support? If so, we'd need more changes, which we can't fit directly into an API that's meant just to do base64 encoding.
> 
> Does JWS or JWK use plain Base64 encoding or does it use Base64 Url encoding? Base64 Url encoding has other subtle differences (the character set has '-' and '_' instead of '+', '/'): https://en.wikipedia.org/wiki/Base64#URL_applications

Good points. This comes from the fact that [JWS explicitly drops the padding characters](https://datatracker.ietf.org/doc/html/rfc7515#appendix-C). It is also base64url encoded but that's easy to swap with a helper function. We can discuss what might be a more elegant solution to this? It's definitely needed for ADU and we would rather not have to add in the padding characters since we would have to memcpy the whole buffer over.

Reply:
A strawman solution is to have new base64url encode and decode APIs with an additional bool parameter for optional/required padding characters emitted or required (depending on encode vs decode). They'd use a new array for the encode and decode character set, and otherwise largely share most of the implementation detail (common layer can be refactored out into helpers, to reduce size). Presumably, adding those extra 256+64 bytes is ok from a code size perspective for the embedded space, and the compiler can optimize out unused code.

That solution is cleaner from an API design perspective, conveys intent correctly, and keeps them single responsibility (also, we want to keep the existing base64 API true to what it does, which fails on invalid input on decode). So, unless there are some concerns with that, I suggest starting with that.

The character swap/replacement logic in an helper that you were thinking of is a decent solution, but it doesn't address the padding concern and I agree doing a memcpy to add the padding can be an issue (removing padding while encoding seems easy). The method though would be slower than always encoding correctly right of the bat (more work to go through the data in the buffer again), but there are runtime cost vs code size trade-offs that we might need to consider here.

How's the data that we need to process being encoded? Would the data being decoded typically get encoded with the use of our Base64 encode APIs or is the expectation that they get encoded from somewhere else?

Reply:
> How's the data that we need to process being encoded? Would the data being decoded typically get encoded with the use of our Base64 encode APIs or is the expectation that they get encoded from somewhere else?

This is coming from the service. So we take in the payload and have to decode it before doing things with the decoded characters.

Reply:
@danewalton would it be ok if the base64url decoder didn't allow padding characters at all?

@LarryOsterman given padding characters are optional in base64url, should the decoder reject any payload that has padding as invalid, or be flexible and ignore them? I was thinking of either letting the caller make the decision about being extra strict (with a param) or being more flexible in what it accepts by default.

If we decide not to have the bool parameter, we can always be strict on the decoder to start with and allow padding characters, if need be, in the future, as that will be non-breaking.

The encoder not producing any padding characters from the get-go, and not giving an option for the caller to do so, seems reasonable.

Reply:
I see no reason to not allow for the padding characters to be there. It doesn't seem like the URL doesn't allow it, but merely gives the option to not have them. I'm not sure we lose any value essentially keeping what I have in this PR, which is that it will assume padding if they are missing, and if the length is `length % 4 == 0`, then it keeps whatever is there.

Reply:
@ahsonkhan Padding characters are optional in base64url encoding. If they're there, the decoder should ignore them. The encoder should *not* produce them. Adding a parameter doesn't help the situation and only complicates the API surface (and as I've mentioned elsewhere, I've yet to encounter a situation where base64url encoded content doesn't have a well-defined length (largely because protocol designers who specify base64url encoding usually ensure that there is *something* in the protocol which can be used to determine the length of the encoded data). 

@danewalton What is important is that base64 decode needs to *reject* base64 encoded data that doesn't have padding characters, and base64url decode needs to allow base64url encoded data which doesn't have padding characters. And of course, base64 encode needs to add padding characters and base64url encode should NOT add padding characters.

Reply:
That makes a lot of sense to me, and I agree. So, exclude the bool parameter from the API in the solution I mentioned above (on both sides).

Thanks for clarifying :)

Reply:
> @danewalton What is important is that base64 decode needs to reject base64 encoded data that doesn't have padding characters, and base64url decode needs to allow base64url encoded data which doesn't have padding characters. And of course, base64 encode needs to add padding characters and base64url encode should NOT add padding characters.

Perfect that sounds good to me